File: 2018_09_03_in_c/two_sum.c
#include <stdio.h>
#include <stdlib.h>
#include "hash_table.h"
#include "two_sum.h"

#define DEFAULT_HT_ARRAY_SZ 2749991
#define MAX_NUMBERS_IN_FILE 1000000


int two_sum(HashTable *ht, signed long long int *numbers,
            int numlen, int t) {
  /* 2-sum algorithm
   *
   * Given a set of integers ints and an integer t, return True if
   * there exist distinct elements x and y in ints such that x + y == t
   *
   * Runs in O(n) time - relies on hash table for O(1) lookups
   */
  signed long long int x, y;
  for (int i=0; i<numlen; i++) {
    x = numbers[i];
    y = t - x;
    if (x != y) {
      if (ht->lookup(ht, y)) {
        printf("Found two-sum for %d (x=%lld, y=%lld)\n", t, x, y);
        return 1;
      }
    }
  }
  return 0;
}

int two_sum_wrapper(const char *filename, signed long int t_low_bnd,
                    signed long int t_up_bnd) {
  FILE *fp = fopen(filename, "r");
  signed long long int numbers[MAX_NUMBERS_IN_FILE];
  signed long long int n = 0;
  int i = 0;
  int lineCount = 0;
  while (fscanf(fp, "%lld,", &n) > 0) {
    numbers[i++] = n;
    lineCount++;
  }

  HashTable *ht = initializeHashTable(DEFAULT_HT_ARRAY_SZ);
  for (i=0; i<lineCount; i++) {
    ht->insert(ht, numbers[i]);
  }

  unsigned int successes = 0;
  for (i=t_low_bnd; i<t_up_bnd+1; i++) {
    if (two_sum(ht, numbers, lineCount, i)) {
      successes++;
    } else {
      fprintf(stderr, "Couldn't find two-sum for %d\n", i);
    }
  }
  fprintf(stderr, "Found two-sums for %d numbers\n", successes);
  return successes;
}

File: 2018_09_03/two_sum.py
from hash_tables import dict_wrapper, hash_table, set_wrapper


def two_sum(ints, t, populated_hash_table):
    """2-sum algorithm.

    Given a set of integers ints and an integer t, return True if
    there exist distinct elements x and y in ints such that x + y == t
    """
    for x in ints:
        y = t - x
        if y != x and populated_hash_table.lookup(y) is True:
            return True
    return False


def do_two_sum_over_range(ints, begin_incl, end_incl, hash_table_class):
    """Return the number of times the two-sum algorithm succeeds
    for a given set of integers and a range of possible t values
    """
    populated_hash_table = hash_table_class()
    for integer in ints:
        populated_hash_table.insert(integer, True)

    successes = 0
    for t in range(begin_incl, end_incl + 1):
        if two_sum(ints, t, populated_hash_table):
            print("Found two-sum for {}".format(t))
            successes += 1
        else:
            print("Couldn't find two-sum for {}".format(t))
    return successes


def main():
    with open("two_sum_data.txt", "r") as file:
        ints = [int(line) for line in file.readlines()]
    result = do_two_sum_over_range(ints, -10000, 10000, set_wrapper)
    print("Result: {}".format(result))
    return result


if __name__ == "__main__":
    main()
